pytorch_lightning/utilities/data.py

```diff
@@ -320,30 +338,44 @@ def _auto_add_worker_init_fn(dataloader: DataLoader, rank: int) -> None:
         dataloader.worker_init_fn = partial(pl_worker_init_function, rank=rank)
 
 
-def _wrap_init(init: Callable) -> Callable:
-    """Wraps the ``__init__`` method of the dataloader in order to enable re-instantiation of custom subclasses of
-    :class:`~torch.utils.data.DataLoader`."""
+def _wrap_dataloader_init(init: Callable) -> Callable:
+    """Wraps the ``__init__`` method of :class:`~torch.utils.data.DataLoader` in order to enable re-instantiation
+    of custom subclasses."""
 
     @functools.wraps(init)
     def wrapper(obj: DataLoader, *args: Any, **kwargs: Any) -> None:
         # We need to inspect `init`, as inspecting `obj.__init__`
         # can lead to inspecting the wrong function with multiple inheritance
         params = inspect.signature(init).parameters
-
         param_names = [
             param.name
             for param in params.values()
             if param.name != "self" and param.kind not in (param.VAR_POSITIONAL, param.VAR_KEYWORD)
         ]
+        param_names = param_names[: len(args)]
+
+        if not hasattr(obj, "__pl_dl_args"):
+            set_arg_names = set()
+            for arg_name, arg_value in chain(zip(param_names, args), kwargs.items()):
+                # Set the value privately if it has not been set yet. This achieves two things:
+                # 1. The argument is the one that was passed in the outermost call
+                #    and not overwritten by a call to `super().__init__()`
+                # 2. The argument is the passed value and does not get overwritten in `__init__()`
+                setattr(obj, "__" + arg_name, arg_value)
+                set_arg_names.add(arg_name)
+            obj.__pl_dl_args = set_arg_names
+
+        if not hasattr(obj, "__dataset"):
+            # We have not found value for dataset yet, but we will find it eventually
+            # because it has to be passed to the original DataLoader
+            dataset_value = None
+            if "dataset" in param_names:
+                dataset_value = args[param_names.index("dataset")]
+            elif "dataset" in kwargs:
+                dataset_value = kwargs["dataset"]
+            if dataset_value is not None:
```

Comment:
Edge case: if the actual value passed is None then this would break: `dataset=None`

You can fix it by duplicating the `setattr` inside the `if`s above

And add a test case for that!